rwrap.1
.TH man 1 "26 June 2018" "1.0" "rwrap man page"

.SH NAME

rwrap \- backup or restore all user profiles on a host

.SH SYNOPSIS

rwrap [-br] [uxv] [-hH] [SHARE] [MOUNTPOINT]

.SH REQUIRED PARAMETERS

.IP -b
Backup user profiles from /Users to /mountpoint/Users/user/date.
.IP -r
Restore user profiles from /mountpoint/Users/user/date to /Users.

.SH OPTIONAL PARAMETERS

.IP -u
Process all users without prompting.
.IP -x
Automatically choose sane excludes (Caches, Trash, etc).
.IP -v
Verbose stream. Useful for debugging.
.IP -h
Display this help and exit.

.SH ARGUMENTS

Takes a share url as an argument, in either standard mount_smbfs form:
//username:password@host/path/to/PARENT/of/Users, or standard mount_nfs form:
host:path/to/PARENT/of/Users, or standard rsync URL form:
//rsync:host:port/path/to/PARENT/of/Users

Takes a mountpoint as an argument: /Volumes/Backups. If a share has been entered
it will be mounted here, otherwise we backup to this location. If no share or
mountpoint is entered you will be prompted to enter them. If your share is a
valid rsync url, there is no need to enter a mountpoint.

.SH PITFALLS & CAVEATS

THE PATH ON THE REMOTE HOST MUST POINT TO PARENT OF Users, NOT Users!

Once mounted the mountpoint should contain the Users directory underneath it.
The script backups to or restores from the following directory layout underneath
the mountpoint: Users/user.name/date, and logs to:
Logs/CopyType/user.name/date.log, where CopyType equals either Backup or
Restore. If this layout does not exist it will be created under the mountpoint.

Unless backing up to a samba share, user, group and permissions will be retained
and each backup will be hard-linked against the last to save space and mimic the
functionality of Time Machine.

https://blog.interlinked.org/tutorials/rsync_time_machine.html

When using Samba, in combination with Active Directory, ownership & permissions
attributes are mangled, so these are not preserved in this use case, making
incremental backups with hard links impossible. This is also the reason it’s
necessary to recursively chown the users directory when restoring from a Samba
share. If possible use this script in conjunction with either an NFS or rsync
server, or a USB drive.

It's also worth pointing out that for the time being preserving extended
attributes is not supported.

.SH EXAMPLES

To backup from /Users to a Samba share mounted at /Volumes/Backups/Users:

backup -b //user:password@hostname/Backups/macOS /Volumes/Backups

To restore from a Samba share mounted at /Volumes/Backups/Users to /Users,
choosing sane defaults:

backup -rx //user:password@hostname/Backups/macOS /Volumes/Backups

To backup from /Users to a rsync server, without prompting for each user:

backup -bu rsync://user@host:port/backups

To restore from /Users to a rsync server, with increased verbosity:

restore -rv rsync://user@host:port/backups

.SH SEE ALSO

rsync(1), rsnapshot(1)

.SH BUGS

Here be dragons!

.SH AUTHOR

Toby Slight (tobyslight@gmail.com)


src/rwrap.py
import argparse
import curses
import os
import pwd
import re
import subprocess
import time

from ui import ask
from ppick import select
os.environ.setdefault('ESCDELAY', '12')  # otherwise it takes an age!

RSYNC_OPTS = [
    '--archive',
    '--human-readable'
]

DEFAULT_EXCLUDES = [
    '--exclude="desktop.ini"',
    '--exclude="*.ost"',
    '--exclude="*.pst"',
    '--exclude="*spotify*"',
    '--exclude="*Spotify*"'
]


def check_dir(path):
    """
    Check if a directory exists and if not try to create it.
    """
    if (not(os.path.isdir(path))):
        try:
            os.makedirs(path)
        except OSError:
            print("Could not create directory.")


def get_excludes(path, hidden):
    """
    Takes a path as an argument and returns a list of child paths that the user
    has selected.
    """
    excludes = curses.wrapper(select, path, hidden)
    if len(excludes) > 0:
        print("\nSelected excludes:\n"+('\n'.join(excludes)))
        if ask("\nAccept and use excludes? "):
            return excludes
        else:
            excludes = get_excludes(path, hidden)
    else:
        if not(ask("\nNo excludes selected. Continue? ")):
            excludes = get_excludes(path, hidden)
    return excludes


def get_users():
    """
    Return a dictionary of users. User name is the key, home directory path is
    the value.
    """
    users = {}
    for p in pwd.getpwall():
        user = p[0]
        home = p[5]
        regex = "^_|admin|daemon|guest|local|nobody|root"
        if not(re.match(regex, user, re.IGNORECASE)):
            users[user] = home

    return users


def get_args():
    """
    Return a list of valid arguments.
    """
    parser = argparse.ArgumentParser(description='\
    Backup or restore users to an rsync server.')
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument("-b", "--backup", action="store_true",
                       help="Backup users to rsync server or path.")
    group.add_argument("-r", "--restore", action="store_true",
                       help="Restore users on from last backup.")
    parser.add_argument("-u", "--users", action="store_true",
                        help="Automate user selection.")
    parser.add_argument("-x", "--excludes", action="store_true",
                        help="Automate excludes selection.")
    group = parser.add_mutually_exclusive_group()
    group.add_argument("-v", "--verbose", action="store_true",
                       help="Increase output verbosity.")
    group.add_argument("-q", "--quiet", action="store_true",
                       help="Run silently.")
    parser.add_argument("url", type=str, help="A valid rsync url.")
    return parser.parse_args()


def copy_skel(date, user, url):
    parent = "/tmp"
    skel = parent + "/Users/" + user + "/" + date
    check_dir(skel)
    cmd = "rsync " + ' '.join(RSYNC_OPTS) + " --quiet " + skel + " " + url
    subprocess.call(cmd, shell=True)


def rwrap(src, dest, automate_excludes):
    hidden = True
    if automate_excludes:
        cmd = "rsync " + " ".join(RSYNC_OPTS) + \
            " ".join(DEFAULT_EXCLUDES) + " " + src + " " + dest
        subprocess.call(cmd, shell=True)
    else:
        excludes = get_excludes(src, hidden)
        # list + set removes duplicates.
        excludes = excludes + list(set(excludes + DEFAULT_EXCLUDES))
        cmd = "rsync " + " ".join(RSYNC_OPTS) + \
            " ".join(excludes) + src + " " + dest
        subprocess.call(cmd, shell=True)


def main():
    users = get_users()
    args = get_args()

    if len(users) > 0:
        print("Users: "+(', '.join(users.keys())))
        for user, home in users.items():
            if args.backup:
                copytype = "backup"
                date = time.strftime("%Y-%m-%d")
                src = home + "/"
                dest = args.url + "/Users/" + user + "/" + date
                copy_skel(date, user, args.url)
            elif args.restore:
                copytype = "restore"
                upath = (args.url + "/Users/" + user)
                date = sorted(os.listdir(upath), reverse=True)[0]
                src = upath + "/" + date + "/"
                dest = home

            if args.users:
                rwrap(src, dest, args.excludes)
            else:
                question = "\n" + copytype.title() + " " + user + "? "
                if ask(question):
                    rwrap(src, dest, args.excludes)
    else:
        print("No users to %s" % copytype)


main()
